beater/otel.go

```diff
@@ -0,0 +1,236 @@
+	otc.observer.cancel()
+	for _, c := range otc.collectors {
+		otc.logger.Infof("Stopping collector %s listening on: %s", c.name(), c.endpoint())
+		c.stop()
+	}
+}
+
+type observer struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	ch     chan error
+}
+
+func newObserver(ctx context.Context) *observer {
+	ctx, cancel := context.WithCancel(ctx)
+	return &observer{ctx: ctx, cancel: cancel, ch: make(chan error)}
+}
+
+func (o *observer) fatal(err error) {
+	o.ch <- err
+}
+
+type jaegerCollector struct {
+	grpcServer *grpc.Server
+	host       string
+	receiver   *jaegerReceiver
+	observer   *observer
+}
+
+func newJaegerCollector(cfg *config.Config, traceConsumer *Consumer, tracer *apm.Tracer, observer *observer) (*jaegerCollector, error) {
+	if cfg.OtelConfig == nil || !cfg.OtelConfig.Jaeger.Enabled {
+		return nil, nil
+	}
+
+	grpcOptions := []grpc.ServerOption{grpc.UnaryInterceptor(apmgrpc.NewUnaryServerInterceptor(
+		apmgrpc.WithRecovery(),
+		apmgrpc.WithTracer(tracer)))}
+	if cfg.OtelConfig.Jaeger.GRPC.TLS != nil {
+		creds, err := credentials.NewServerTLSFromFile(cfg.OtelConfig.Jaeger.GRPC.TLS.Certificate, cfg.OtelConfig.Jaeger.GRPC.TLS.Key)
+		if err != nil {
+			return nil, err
+		}
+		grpcOptions = append(grpcOptions, grpc.Creds(creds))
+	}
+	grpcServer := grpc.NewServer(grpcOptions...)
+
+	return &jaegerCollector{
+		receiver:   &jaegerReceiver{traceConsumer: traceConsumer},
+		grpcServer: grpcServer,
+		host:       cfg.OtelConfig.Jaeger.GRPC.Host,
+		observer:   observer,
+	}, nil
+}
+
+func (jc *jaegerCollector) start() error {
+	api_v2.RegisterCollectorServiceServer(jc.grpcServer, jc.receiver)
+	listener, err := net.Listen(networkTCP, jc.host)
+	if err != nil {
+		return err
+	}
+	go func() {
+		if err := jc.grpcServer.Serve(listener); err != nil {
+			jc.observer.fatal(err)
+		}
+	}()
+	return nil
+}
+
+func (jc *jaegerCollector) stop() {
+	jc.grpcServer.GracefulStop()
+}
+
+func (jc *jaegerCollector) name() string {
+	return jaegerType
+}
+
+func (jc *jaegerCollector) endpoint() string {
+	return jc.host
+}
+
+type jaegerReceiver struct {
+	traceConsumer *Consumer
+}
+
+// PostSpans implements the api_v2/collector.proto. It converts spans received via Jaeger Proto batch to open-telemetry
+// TraceData and passes them on to the internal Consumer taking care of converting into Elastic APM format.
+// The implementation of the protobuf contract is based on the open-telemetry implementation at
+// https://github.com/open-telemetry/opentelemetry-collector/tree/master/receiver/jaegerreceiver
+func (jr *jaegerReceiver) PostSpans(ctx context.Context, r *api_v2.PostSpansRequest) (*api_v2.PostSpansResponse, error) {
+	ctxWithReceiverName := observability.ContextWithReceiverName(ctx, jaegerReceiverName)
+	spansCount := len(r.Batch.Spans)
+	traceData, err := jaeger.ProtoBatchToOCProto(r.Batch)
+	if err != nil {
+		observability.RecordMetricsForTraceReceiver(ctxWithReceiverName, spansCount, spansCount)
```

Comment:
This method updates some stats in the opentelemetry-collector: https://github.com/open-telemetry/opentelemetry-collector/blob/master/observability/observability.go

Can we replace those calls with the usual ones from `libbeat/monitoring`? That could be done in a followup, so maybe just remove the references to the `observability` package for now and add a TODO?